File: training-data/objectivec/StatePatternDemo.m
#import <Foundation/Foundation.h>

@class TrafficLight;

@protocol LightState <NSObject>
- (void)handle:(TrafficLight *)light;
- (NSString *)name;
@end

@interface TrafficLight : NSObject
@property (nonatomic, strong) id<LightState> state;
- (void)change;
@end

@implementation TrafficLight
- (void)change {
    [self.state handle:self];
}
@end

@interface RedState : NSObject <LightState>
@end
@interface GreenState : NSObject <LightState>
@end
@interface YellowState : NSObject <LightState>
@end

@implementation RedState
- (void)handle:(TrafficLight *)light { light.state = [[GreenState alloc] init]; }
- (NSString *)name { return @"Red"; }
@end

@implementation GreenState
- (void)handle:(TrafficLight *)light { light.state = [[YellowState alloc] init]; }
- (NSString *)name { return @"Green"; }
@end

@implementation YellowState
- (void)handle:(TrafficLight *)light { light.state = [[RedState alloc] init]; }
- (NSString *)name { return @"Yellow"; }
@end

int main(int argc, const char *argv[]) {
    @autoreleasepool {
        TrafficLight *light = [[TrafficLight alloc] init];
        light.state = [[RedState alloc] init];
        for (NSInteger i = 0; i < 4; i++) {
            NSLog(@"%@", [light.state name]);
            [light change];
        }
    }
    return 0;
}

File: training-data/csharp/StatePatternDemo.cs
using System;

interface ITrafficLightState
{
    ITrafficLightState Next();
    string Name { get; }
}

class RedState : ITrafficLightState
{
    public string Name => "Red";
    public ITrafficLightState Next() => new GreenState();
}

class GreenState : ITrafficLightState
{
    public string Name => "Green";
    public ITrafficLightState Next() => new YellowState();
}

class YellowState : ITrafficLightState
{
    public string Name => "Yellow";
    public ITrafficLightState Next() => new RedState();
}

class TrafficLight
{
    private ITrafficLightState state = new RedState();

    public void Advance() => state = state.Next();
    public string CurrentState => state.Name;
}

class StatePatternDemo
{
    static void Main()
    {
        var light = new TrafficLight();
        for (int i = 0; i < 5; i++)
        {
            Console.WriteLine(light.CurrentState);
            light.Advance();
        }
    }
}
